Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,516 @@
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request)
+        switch self.getAction(for: key, on: eventLoop) {
+        case .useExisting(let provider):
+            return provider.getConnection(preference: preference)
+
+        case .make(completing: let providerPromise):
+            // If no connection provider exists for the given key, create a new one
+            let selectedEventLoop = providerPromise.futureResult.eventLoop
+            var bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: selectedEventLoop, host: request.host, port: request.port, configuration: self.configuration) { _ in
+                self.loopGroup.next().makeSucceededFuture(())
+            }
+
+            if let timeout = resolve(timeout: self.configuration.timeout.connect, deadline: deadline) {
+                bootstrap = bootstrap.connectTimeout(timeout)
+            }
+
+            let address = HTTPClient.resolveAddress(host: request.host, port: request.port, proxy: self.configuration.proxy)
+
+            return bootstrap.connect(host: address.host, port: address.port).flatMap { channel in
+                channel.pipeline.addSSLHandlerIfNeeded(for: key, tlsConfiguration: self.configuration.tlsConfiguration).flatMap {
+                    channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+                }.flatMap {
+                    let provider = ConnectionProvider.http1(HTTP1ConnectionProvider(group: self.loopGroup, key: key, configuration: self.configuration, initialConnection: Connection(key: key, channel: channel, parentPool: self), parentPool: self))
+                    providerPromise.succeed(provider)
+                    return provider.getConnection(preference: preference)
+                }
+            }
+        }
+    }
+
+    func release(_ connection: Connection) {
+        self.lock.withLock {
+            if let provider = self._connectionProviders[connection.key] {
+                provider.release(connection: connection)
+            }
+        }
+    }
+
+    func syncClose(requiresCleanClose: Bool) throws {
```

Comment:
@Lukasa Should this throw or should this fail using `preconditionFailure`? AFAICT the only reason why this would fail would be an internal bug, so does this actually make sense to make a new `HTTPClientError` case for it?

Reply:
internal bugs should use `preconditionFailure` instead of errors.

Reply:
@weissi Thanks